### ISSUE TITLE
LoRA Training on the GUI - Pause after each epoch #2

### ISSUE THREAD

**Issue by dopf-26** [ace-step/ACE-Step-1.5#216]
Sadly commit https://github.com/ace-step/ACE-Step-1.5/pull/177 has not fixed the issue. Training still pauses like before on #358325e.

```
2026-02-06 09:51:18.564 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 32s] Step 70: Epoch 3/1000, Step 70, Loss: 0.7975
2026-02-06 09:51:24.076 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 37s] Step 80: Epoch 3/1000, Step 80, Loss: 0.9445
2026-02-06 09:51:29.943 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 43s] Step 90: Epoch 3/1000, Step 90, Loss: 0.8529
2026-02-06 09:51:34.954 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 48s] Step 96: ✅ Epoch 3/1000 in 37.8s, Loss: 0.8581
Loaded configuration from C:\Users\xxx\GenAI\Audio\ACE-Step-1.5_portable\.env
Loaded configuration from C:\Users\xxx\GenAI\Audio\ACE-Step-1.5_portable\.env
Loaded configuration from C:\Users\xxx\GenAI\Audio\ACE-Step-1.5_portable\.env
Loaded configuration from C:\Users\xxx\GenAI\Audio\ACE-Step-1.5_portable\.env
Skipping import of cpp extensions due to incompatible torch version 2.7.1+cu128 for torchao version 0.15.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.7.1+cu128 for torchao version 0.15.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.7.1+cu128 for torchao version 0.15.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
Skipping import of cpp extensions due to incompatible torch version 2.7.1+cu128 for torchao version 0.15.0             Please see https://github.com/pytorch/ao/issues/2919 for more info
W0206 09:51:40.351000 12800 Lib\site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
W0206 09:51:40.362000 13152 Lib\site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
W0206 09:51:40.362000 4408 Lib\site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
W0206 09:51:40.425000 10996 Lib\site-packages\torch\distributed\elastic\multiprocessing\redirects.py:29] NOTE: Redirects are currently not supported in Windows or MacOs.
```

I asked gemini for a reason, If i do the following edits I can keep training running without any kind of pauses:

Change in acestep/training/configs.py:
- save_every_n_epochs: int = 50
- num_workers: int = 0 #this is the main culprit
- log_every_n_steps: int = 50

Change in acestep/training/trainer.py:
- Add this after the imports: "torch.set_float32_matmul_precision('high')"
- accelerator="cuda"
- and right below: strategy="auto"

This is the same dataset, same settings after changing the above lines. Getting 2x speed and no more errors:
```
2026-02-06 09:58:27.872 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1s] Step 0: 🚀 Starting training (precision: bf16-mixed)...
2026-02-06 09:58:27.879 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1s] Step 0: 🎯 Training 44,040,192 parameters
2026-02-06 09:58:48.159 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [21s] Step 32: ✅ Epoch 1/1000 in 20.2s, Loss: 1.2864
2026-02-06 09:58:58.429 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [32s] Step 50: Epoch 2/1000, Step 50, Loss: 1.0029
2026-02-06 09:59:06.188 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [39s] Step 64: ✅ Epoch 2/1000 in 18.0s, Loss: 1.0008
2026-02-06 09:59:24.441 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [58s] Step 96: ✅ Epoch 3/1000 in 18.2s, Loss: 0.8645
2026-02-06 09:59:26.697 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 0s] Step 100: Epoch 4/1000, Step 100, Loss: 0.7134
2026-02-06 09:59:42.491 | INFO     | acestep.gradio_ui.events.training_handlers:start_training:585 - [1m 16s] Step 128: ✅ Epoch 4/1000 in 18.0s, Loss: 0.8032
```

It also got rid of the pesky torchao error message, this one now only shows at startup. Torchao can be downgraded to 0.12.0 to get rid of the message alltogether, but I dont know how that translates for performance or other errors.

Sadly Im not a coder and too stupid to see what else I am affecting with these changes, but you can also use AdamW8bit with bitsandbytes instead of AdamW to speedup training further.

Edit: With the above changes, plus "torch.backends.cudnn.benchmark = True", enabled Torch Compile and AdamW8Bit i now get about 16.6s per epoch and could bump up the batch size from 3 to 5. So insane speed gains here if the training actually works, which I'll have to test.

**Comment by Meliauo** [comment]
https://discord.com/channels/1369256267645849741/1469143590176297072

You can take a look at this, but it hasn't been implemented yet.

**Comment by dopf-26** [comment]
> https://discord.com/channels/1369256267645849741/1469143590176297072
> 
> You can take a look at this, but it hasn't been implemented yet.

It's a bit cryptic, but hopefully Moonspell will do a proper pull request. I could try as well but someone with more knowledge would have to take a look at the changes and check if it messes anything up for linux / macOS and so on.

**Comment by maizhouzi** [comment]
Thanks @Chilluminati91 for the initial suggestions! Inspired by your post, I did some deep diving into the memory issues on my 5070 Ti (16GB VRAM).

Since I'm not a professional coder, I used Claude Opus 4.6 to overhaul the training pipeline based on the bottlenecks I encountered. The results are quite massive:

Performance Gains (Windows @ 5070 Ti 16GB):

Before: Batch Size 1 occupied 16GB VRAM + 2GB Shared VRAM (extremely slow due to off-device swapping).
After: Batch Size 6 now only uses ~9GB VRAM, resulting in a ~40x speedup compared to the previous bottlenecked state.
Key Optimizations Implemented:
Memory Hygiene: Offloading unused components (Encoder/VAE/Text Encoder) to CPU during the training loop.
Optimizer: Switched to AdamW8bit (bitsandbytes) to save weight-related memory.
Preprocessing: Implemented Tiled VAE Encoding and strict garbage collection (gc.collect() + cuda.empty_cache()) to prevent VRAM spikes during tensor generation.
Stability:
Fixed the specific Windows multiprocess stall by forcing num_workers=0.
Added a script-level fix to disable CMD QuickEdit Mode (preventing accidental hangs).
Implemented auto-detection of latest checkpoints and Emergency Save on KeyboardInterrupt.
Efficiency: Added "Skip existing samples" logic to support resuming preprocessing.
The training loss looks stable so far. I’ve uploaded these changes to my fork for anyone struggling with 16GB VRAM limits on Windows.

Fork Link:
https://github.com/maizhouzi/ACE-Step-1.5/tree/training-optimizations

**Comment by Maelstrom2014** [comment]
Great! I Alse recomend to use acestep-v15-sft  for training lora.
